dbms/src/Flash/Pipeline/Schedule/TaskQueue.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Pipeline/Schedule/TaskQueue.h>
+#include <assert.h>
+#include <common/likely.h>
+
+namespace DB
+{
+bool FIFOTaskQueue::take(TaskPtr & task)
+{
+    assert(!task);
+    {
+        std::unique_lock lock(mu);
+        while (true)
+        {
+            if (unlikely(is_closed))
+                return false;
+            if (!task_queue.empty())
+                break;
+            cv.wait(lock);
+        }
+
+        task = std::move(task_queue.front());
+        task_queue.pop_front();
+    }
+    assert(task);
+    return true;
+}
+
+bool FIFOTaskQueue::empty()
+{
+    std::lock_guard lock(mu);
+    return task_queue.empty();
+}
+
+void FIFOTaskQueue::close()
+{
+    {
```

Comment:
Shall we assert that `!is_closed` is hold?

Reply:
I don't think it matters, the difference is whether `close` can be called more than once.